.changeset/dry-papers-warn.md

```diff
@@ -0,0 +1,31 @@
+---
+"@fluidframework/container-runtime": major
+"@fluidframework/container-runtime-definitions": minor
+"@fluidframework/datastore": major
+"@fluidframework/runtime-definitions": minor
+"@fluidframework/runtime-utils": minor
+"@fluid-private/test-end-to-end-tests": minor
+"@fluidframework/test-utils": minor
+---
+
+Implement ID compression for DataStores & DDSs
+
+
+# Key changes
+1. A new API IContainerRuntimeBase.generateDocumentUniqueId() is exposed. This API will opportunistically generate IDs in short format (non-negative numbers). If it can't achieve that, it will return a uuid string. It can be leveraged anywhere in container where container unique IDs are required. I.e. any place that uses uuid() and stores data in container is likely candidate to start leveraging this API.
+2. Data store internal IDs (IDs that are auto generated by FF system) will opportunistically be generated in shorter form. Data stores created in detached container will always have short IDs, data stores created in attached container will opportunistically be short (by using newly added IContainerRuntimeBase.generateDocumentUniqueId() capability)
```

Comment:
how did you accomplish detached container ID generation always generating short IDs? this isn't a feature of the compressor (though we could make it aware of the detached state, theoretically)

Reply:
There are two different layers - container runtime (assigning IDs to data stores) and data store (assigning IDs to DDSs). In both cases it's parent-child relationship where child knows that it's visibility is local (not global, i.e. only one client has access to it) - it's detached container for data stores and detached data store for DDS.
In both cases I simply assign them numbers 1, 2, 3, etc. based on number of objects under parent (as issuing ID === inserting into the list).

Then I take those two non-negative spaces (detached numbers, and numbers from ID compressor) and mix them up into single namespace by doing the following:

ID = 2 * detachedID
ID = 2 * attachedID + 1.

That's it.

Reply:
Ah, I'd wondered if you were going to do something sneaky with [fibonacci coding](https://en.wikipedia.org/wiki/Fibonacci_coding) or something :)
This seems reasonable, although a bit opaque.